db/db_impl/compacted_db_impl.cc

```diff
@@ -60,37 +79,62 @@ Status CompactedDBImpl::Get(const ReadOptions& options, ColumnFamilyHandle*,
   return Status::NotFound();
 }
 
-std::vector<Status> CompactedDBImpl::MultiGet(const ReadOptions& options,
-    const std::vector<ColumnFamilyHandle*>&,
+std::vector<Status> CompactedDBImpl::MultiGet(
+    const ReadOptions& options, const std::vector<ColumnFamilyHandle*>&,
     const std::vector<Slice>& keys, std::vector<std::string>* values) {
+  return MultiGet(options, keys, values, /*timestamps*/ nullptr);
+}
+
+std::vector<Status> CompactedDBImpl::MultiGet(
+    const ReadOptions& options, const std::vector<ColumnFamilyHandle*>&,
+    const std::vector<Slice>& keys, std::vector<std::string>* values,
+    std::vector<std::string>* timestamps) {
   assert(user_comparator_);
-  if (user_comparator_->timestamp_size() || options.timestamp) {
-    // TODO: support timestamp
-    return std::vector<Status>(keys.size(), Status::NotSupported());
+  size_t num_keys = keys.size();
+
+  if (options.timestamp) {
+    Status s =
+        FailIfTsSizesMismatch(DefaultColumnFamily(), *(options.timestamp));
+    if (!s.ok()) {
+      return std::vector<Status>(num_keys, s);
+    }
+  } else {
+    Status s = FailIfCfHasTs(DefaultColumnFamily());
+    if (!s.ok()) {
+      return std::vector<Status>(num_keys, s);
+    }
   }
+
+  GetWithTimestampReadCallback read_cb(kMaxSequenceNumber);
   autovector<TableReader*, 16> reader_list;
   for (const auto& key : keys) {
-    const FdWithKeyRange& f = files_.files[FindFile(key)];
-    if (user_comparator_->Compare(key, ExtractUserKey(f.smallest_key)) < 0) {
+    LookupKey lkey(key, kMaxSequenceNumber, options.timestamp);
+    const FdWithKeyRange& f = files_.files[FindFile(lkey.user_key())];
+    if (user_comparator_->Compare(lkey.internal_key(),
```

Comment:
I think here the check should be
```cpp
if (user_comparator_->CompareWithoutTimestamp(key, /*a_has_ts=*/false, ExtractUserKeyAndStripTimestamp(f.smallest_key), /*b_has_ts=*/false)) {
  reader_list.push_back(nullptr);
}
```
An example of why use `CompareWithoutTimestamp()`. Suppose you are searching for 'a' and want to return the most recent version smaller than or equal to timestamp 200.
```
[('a', 100)   ('a', 90)]
```
In `user_comparator_->Compare()`, both keys should be user keys with or without timestamp. Use the above overloaded version to specify whether timestamp is included or not.
Furthermore, add a test to cover this case.

Reply:
Good catch. I think the previous code happens to work just because below sanity check logic in the beginning of the function kind of guarantees "keys should be user keys with or without timestamp".
```
  if (options.timestamp) {
     ...
  } else {
    ...
  }
```

Reply:
Thank you for the detailed offline explanation of the purpose of this change. My previous comment was incorrectly assuming internal_key is user_key + timestamp.